File: data_structure_cpp/9002.cpp
////和上一题的不同
////这里是处理数字
////需要使用Dijkstra的双栈算法
////算法思路：一个栈用来放数字，一个栈用来放符号
//#include <stack>
//#include <iostream>
//#include<string>
//#include <cctype> //为了使用isdigit()来判断是不是数字'
//using namespace std;
//int getPrecedence(char op)
//{
//	//和上面一题这个部分一样
//	if (op == '+' || op == '-') {
//		return 1;
//	}
//	if (op == '*' || op == '/') {
//		return 2;
//	}
//	return 0; // '(' 和其他字符
//}
//
////创建一个辅助函数，用来处理一次运算
//void applyOp(std::stack<long long >& values, std::stack<char>& ops) {
//
//	//获取栈顶的运算符
//	char op = ops.top();
//	ops.pop();
//
//	//获取栈顶的数字
//	long long b = values.top();
//	values.pop();
//	long long a = values.top();
//	values.pop();
//
//	//需要注意的是运算顺序是 a ops b
//	switch (op) {
//	case '+':
//		values.push(a + b);
//		break;
//	case '-':
//		values.push(a - b);
//		break;
//	case '*':
//		values.push(a * b);
//		break;
//	case '/':
//		values.push(a / b);
//		break;
//	}
//
//
//}
//
////主计算函数
//
//long long evaluate(string expression) {
//	std::stack<long long>vals;
//	std::stack<char>ops;
//
//	for (int i = 0; i < expression.size(); i++) {
//		char c = expression[i];
//
//		//如果遇到数字，还需判断要把数字连接起来处理
//		if (isdigit(c)) {
//			long long num = 0;
//			while (i < expression.size() && isdigit(expression[i])) {
//				//就是发现当前的数字后面还有数字，首先是肯定要变大10倍
//				//然后就是把数字加上
//				num = num * 10 + expression[i] - '0';
//				i++;
//			}
//			vals.push(num);
//			//注意这里循环会+1
//			i--;//所以要-1
//		}
//
//		//处理字符
//		else if (c == '(') {
//			//左括号，直接入栈
//			ops.push(c);
//		}
//		else if (c == ')') {
//			while (!ops.empty() && ops.top() != '(') {
//				applyOp(vals, ops);
//				
//			}
//			ops.pop();
//		}
//		else if (c == '+' || c == '-' || c == '*' || c == '/') {
//			while (!ops.empty() && getPrecedence(ops.top()) >= getPrecedence(c)) {
//			applyOp(vals, ops);
//           
//			}
//			ops.push(c);
//		}
//		else if (c == '#') {
//			//遇到#，就计算所有运算符
//			while (!ops.empty()) {
//				applyOp(vals, ops);
//			}
//			break;
//		}
//	}
//	return vals.top();
//}
//int main() {
//	std::ios_base::sync_with_stdio(false);
//	std::cin.tie(NULL);
//	std::cout.tie(NULL);
//
//	int T;
//	std::cin >> T;
//	while (T--) {
//		std::string expression;
//		std::cin >> expression;
//		std::cout << evaluate(expression) << std::endl;
//	}
//	return 0;
//}

File: data_structure_python/9002.py
import sys
def getPrecedence(c):
    #获取运算符的优先级
    if c == '+' or c == '-':
        return 1
    if c == '*' or c == '/':
        return 2
    return 0

#构造一个辅助函数用来进行数字和运算
def applyOp(vals,ops):
    """
    辅助函数：执行一次运算
    从 ops 弹出一个操作符, 从 vals 弹出两个数字
    计算结果, 再压回 vals
    """
    op = ops.pop()
    val2 = vals.pop()
    val1 = vals.pop()
    
    #注意这里的运算是vals1 abs val2
    if op == '+':
        vals.append(val1 + val2)
    elif op == '-':
        vals.append(val1 - val2)
    elif op == '*':
        vals.append(val1 * val2)
    elif op == '/':
        vals.append(int(val1 / val2))

def evaluate(expression):
    ops =[]
    vals = []
    
    i = 0
    while i < len(expression):
        c = expression[i]
        
        #当获取到的是数字
        if c.isdigit():
            num_str=''
            while i < len(expression) and expression[i].isdigit():
                num_str += expression[i]
                i += 1
            vals.append(int(num_str))
            i -= 1
        elif c=='(':
            ops.append(c)
        elif c==')':
            while ops[-1]!='(' and ops:
                applyOp(vals,ops)
            ops.pop()
        elif c in ['+', '-', '*', '/']:
            while ops and getPrecedence(c) <= getPrecedence(ops[-1]):
                applyOp(vals,ops)
            ops.append(c)
            
        elif c=='#':
           while ops:
                applyOp(vals,ops)
           break
        i+= 1
    return vals.pop()
            
# def main():
#     """主函数，处理输入输出"""
#     try:
#         T_str = sys.stdin.readline()
#         if not T_str:
#             return  # 如果输入为空，则退出
        
#         T = int(T_str.strip())

#         for _ in range(T):
#             line = sys.stdin.readline().strip() # 读取一行并去掉换行符
#             if line:
#                 result = evaluate(line)
#                 print(result)
                
#     except EOFError:
#         pass
#     except Exception as e:
#         # 在OJ中，最好不要打印任何多余信息
#         # print(f"An error occurred: {e}", file=sys.stderr)
#         pass

# if __name__ == "__main__":
#     main()